File: pycon/prolog/puzzle.pl
% board size
on_board(X, Y) :-
    X >= 1, X =< 8,
    Y >= 1, Y =< 8.

% Define possible knight move offsets
knight_offset(2, 1).
knight_offset(1, 2).
knight_offset(-1, 2).
knight_offset(-2, 1).
knight_offset(-2, -1).
knight_offset(-1, -2).
knight_offset(1, -2).
knight_offset(2, -1).

% knight_move(X1, Y1, X2, Y2) means a 
% knight at (X1,Y1) can move to (X2,Y2)
knight_move(X1, Y1, X2, Y2) :-
    knight_offset(DX, DY),
    X2 is X1 + DX,
    Y2 is Y1 + DY,
    on_board(X2, Y2).

File: pycon/prolog/puzzle.py
#horse moves in chess board
from pyswip import Prolog
prolog = Prolog()
prolog.consult("puzzle.pl")
# the hourse is at (x,y) postion
x, y = 8, 8
print(f"Knight at ({x},{y}) can move to:")
for result in prolog.query(f"knight_move({x}, {y}, X2, Y2)"):
    print(f"→ ({result['X2']}, {result['Y2']})")
